compiler/rustc_target/src/spec/mod.rs

```diff
@@ -1385,6 +1385,20 @@ pub enum FramePointer {
     MayOmit,
 }
 
+impl FramePointer {
+    /// It is intended that the "force frame pointer" transition is "one way"
+    /// so this convenience assures such if used
+    #[inline]
+    pub fn ratchet(&mut self, rhs: FramePointer) -> FramePointer {
+        *self = match (*self, rhs) {
+            (FramePointer::Always, _) | (_, FramePointer::Always) => FramePointer::Always,
+            (FramePointer::NonLeaf, _) | (_, FramePointer::NonLeaf) => FramePointer::NonLeaf,
+            _ => FramePointer::MayOmit,
+        };
+        *self
+    }
```

Comment:
This seems a bit over-complex for the one use it has.

Reply:
Its real purpose is to document that frame pointer forcing should go up and not down. I also originally used it in the parse, I suppose that got lost at some point.

Reply:
Are "up" and "down" defined for this? I assume "up" means "more frame pointers". But I don't think that's really what we want - we want it to default to whatever the spec sets, and allow it to be overridden by the command line. If the spec says "always" but the command line is "may omit" (or always or non-leaf) then that should always take preference.

Reply:
https://github.com/rust-lang/rust/blob/9e7aff794539aa040362f4424eb29207449ffce0/src/doc/rustc/src/codegen-options/index.md?plain=1#L157-L165

Reply:
Removing frame pointers can break the standard library's runtime behaviors. A given target can require frame pointers and simply stop working correctly (e.g. backtraces or unwinding may stop working) if software is compiled otherwise. While not everyone values backtraces much (else this compilation option would never have come about), breaking the unwinder seems... bad.

Reply:
This is an advanced option for people who know what they're doing. I think it's important that it do exactly what it claims to do. Obviously it shouldn't generate code that's so broken that it's non-viable, but to the greatest extent possible it should allow overriding the default setting for that target. There are lots of cases where backtraces/unwinding might be considered secondary to whatever goal disabling frame pointers is trying to achieve and are therefore expendible.

>  breaking the unwinder seems... bad.

That's moot if you're using panic=abort. And in many (most? all?) cases unwinding does not rely on frame pointers.

Reply:
On macOS the arm64 ABI mandates the existence of frame pointers. On x86_64 frame pointers are also mandated with panic=unwind: https://github.com/rust-lang/rust/issues/74640

On Windows breaking the unwinder will cause problems even with panic=abort as the OS throws a SEH exception whenever the stack needs to grow. This indeed doesn't depend on frame pointers though.

Reply:
@jsgf It currently does do what it says it does, per the docs. Changing the behavior the way you describe is a much more significant change and one I do not think we should try to smuggle into a much smaller proposal to add a flag.

Reply:
Right now "no" means "MayOmit", which leaves it up to llvm to decide how to behave - this is fine. It still allows the target spec to be overridden.

The way `ratchet` is implemented means that the spec always takes precedence over the command line option if its "stronger". I want to be able to override this. It's also a change of behaviour compared to the current code. Right now, the only thing that always forces frame pointers is using mcount.

Reply:
@jsgf It does not amount to a change in behavior! *Please*, reread the function, the version before I touched it:

https://github.com/rust-lang/rust/blob/6e5f1dacf3a670cf03111039a032c765355839aa/compiler/rustc_codegen_llvm/src/attributes.rs#L106-L120

On line 107, we get this code which mutably declares `fp`:
https://github.com/rust-lang/rust/blob/6e5f1dacf3a670cf03111039a032c765355839aa/compiler/rustc_codegen_llvm/src/attributes.rs#L107

That's the Session's target, which is derived from the spec. We get the field `frame_pointer` from the spec, right there. That is the spec's value. That means it is what we default to if no other code changes it, yes?

On line 111, we get this code:
https://github.com/rust-lang/rust/blob/6e5f1dacf3a670cf03111039a032c765355839aa/compiler/rustc_codegen_llvm/src/attributes.rs#L111

There are two conditions there. One is `mcount`. The other is *also* from the session. It's the session's parsed configuration value for the `force_frame_pointers` option. It is `Some(true)` if you set `-Cforce-frame-pointers=yes` (or any of the other positive values).

These conditions are joined by `||`, a short-circuiting `or`, so only one must be `true` in order for us to reach the next line:
https://github.com/rust-lang/rust/blob/6e5f1dacf3a670cf03111039a032c765355839aa/compiler/rustc_codegen_llvm/src/attributes.rs#L111
...where we set `FramePointer::Always`.

That's the only possible change. The fact that we accept `-Cforce-frame-pointers=no` on the command line is a cruel trick, in effect. It won't actually do anything.

The "target" does in fact take precedence.

Reply:
_Perhaps_ some subtlety has amounted to a change *elsewhere*... we're modifying code in multiple places, after all... but I believe the `frame_pointer_type_attr`, the one that combines the various configuration options into a single "set frame pointers?" value, continues to behave exactly as it did before. For all the currently-settable values that are not being introduced in this PR, anyways. I believe the behavior you dislike is the current behavior.

I believe the code, unfortunately, is now clearer as to how it functions, and no longer misleads you as to what happens. Perhaps that has led you to dislike it? That seems a reasonable response, and I am not trying to persuade you that your requested behavior is unreasonable. I _am_ trying to persuade you that your request is _not what is implemented currently_, and that there are at least _reasons to make it a separate process_.

I think it amounts to promising a new behavior, basically, which as I understand it requires a T-compiler FCP? Again, "that would require more process" is a reason I went for drafting the PR I currently did.

The _new_ thing is the `-Cforce-frame-pointers=non-leaf` setting. I implemented it the way I have because this causes it to align _notionally_ with the way the code currently works: it takes the highest value between the one in the spec, the one given on the command line, and the one required by `instrument_mcount`.

Reply:
Yeah sorry I re-read the original and I agree that spec overrides. Well, I guess we can still use json specs if we really need to.

Reply:
@jsgf I notice your description of the "current behavior" accurately describes the implemented behavior in https://github.com/rust-lang/rust/pull/125127 which, as I mention, unless I am horribly wrong, is different from the current behavior of `frame_pointer_type_attr` in the compiler. I chose not to implement things that way for all the reasons I've already described. Namely: it introduces new behavior that is not how the code currently works.

Please take this line of inquiry to a new *and unrelated* issue, PR, etc.

Reply:
Thank you.